File: Programming Skills/Repeated Substring Pattern/repeatedSubstringPattern.swift
class Solution {
    func repeatedSubstringPattern(_ s: String) -> Bool {
        let mergedStr = s + s
        let newStr = mergedStr.dropFirst().dropLast()
        return newStr.contains(s)
    }
}

class Solution2 {
    func repeatedSubstringPattern(_ s: String) -> Bool {
        let strLen = s.count
        if strLen == 1 { return false }

        for len in 1...strLen / 2 where strLen % len == 0 {
            if s == String(repeating: String(s.prefix(len)), count: strLen / len) {
                return true
            }
        }

        return false
    }
}

File: Programming Skills/Repeated Substring Pattern/repeatedSubstringPattern.c
bool repeatedSubstringPattern(char* s) {
  int strLen = strlen(s);
  
  if (strLen == 1) return false;

  for (int len = 1; len <= strLen / 2; len++) {
    if (strLen % len != 0) continue;

    for (int i = len; i < strLen; i++) {
        if (s[i] != s[i % len]) break;
        if (i == strLen - 1) return true;
    }
  }

  return false;
}